### PR TITLE
Add "Image style" controller

### PR DESCRIPTION
**Pull request by amitaibu**
When passing a special URL, I'd like `ImageMagick` to resize/ crop the image, and save a copy for later re-use

### CODE REVIEW

File: Web/Controller/ImageStyle.hs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                renderFile (cs storedFile.path) "application/jpg"
+
+
+extractDirectoryAndUUID :: Text -> (Text, Text)
+extractDirectoryAndUUID inputText = do
+    case reverse parts of
+        uuid : pathSegments -> (T.intercalate "/" (reverse pathSegments), uuid)
+        _ -> ("", "")
+    where
+        -- @todo: How to get rid of the hardcoded "http://localhost:8000/static" prefix?
```
Comment (amitaibu) [inline Web/Controller/ImageStyle.hs:45]: How to deal with that?

File: Web/Controller/ImageStyle.hs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                -- Image style found.
+                renderFile (cs $ "static/" <> imageStylePath) "application/jpg"
+            else do
+                -- Image style not found, so create it.
+                let options :: StoreFileOptions = def
+                        { directory = imageStylePathDirectory
+                        , preprocess = applyImageMagick "jpg" ["-resize", cs size <> "^", "-gravity", "north", "-crop", cs size, "-quality", "85%", "-strip"]
+                        }
+
+                storedFile <- storeFileFromPath (cs $ "static/" <> originalImageDirectory <> "/" <> uuid) options
```
Comment (amitaibu) [inline Web/Controller/ImageStyle.hs:34]: Need to remove the  hardcoded `static/`, but the bigger issue is this

image

File: Web/Controller/ImageStyle.hs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                renderFile (cs storedFile.path) "application/jpg"
+
+
+extractDirectoryAndUUID :: Text -> (Text, Text)
+extractDirectoryAndUUID inputText = do
+    case reverse parts of
+        uuid : pathSegments -> (T.intercalate "/" (reverse pathSegments), uuid)
+        _ -> ("", "")
+    where
+        -- @todo: How to get rid of the hardcoded "http://localhost:8000/" prefix?
```
Comment (amitaibu) [inline Web/Controller/ImageStyle.hs:45]: Need to find how to get the base URL

File: Web/Controller/ImageStyle.hs (hunk 2 above)
Comment (amitaibu) [reply Web/Controller/ImageStyle.hs:34]: The error is triggered when we have the `preprocess` from above

Comment (amitaibu) [reply Web/Controller/ImageStyle.hs:33]: I've fixed the imageMagick params https://github.com/amitaibu/ihp-landing-page/pull/15/commits/5c0c86e2a7b913d804c663ccc843be54d6be56e5

**Comment by amitaibu** [comment]
image

File: Web/Controller/ImageStyle.hs (hunk 2 above)
Comment (amitaibu) [reply Web/Controller/ImageStyle.hs:32]: Next problem, how to set the filename on `storeFileFromPath` -- right now it creates duplicate files on each reload 

image

Comment (amitaibu) [reply Web/Controller/ImageStyle.hs:32]: https://github.com/digitallyinduced/ihp/blob/248cb16b361adb4e8838b33a496f7c2be09541de/IHP/FileStorage/ControllerFunctions.hs#L181

@mpscholten is there a good reason for why the filename is `""`? Should I write a PR to change it to take the same filename as the original file?

Comment (amitaibu) [reply Web/Controller/ImageStyle.hs:32]: https://github.com/digitallyinduced/ihp/pull/1686

File: Web/Controller/ImageStyle.hs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+instance Controller ImageStyleController where
+    action RenderImageStyleAction { width, height, originalImagePath } = do
+        -- Get the original image directory and UUID from the path.
+        let (originalImageDirectory, uuid) = extractDirectoryAndUUID originalImagePath
+
+        let size = show width <> "x" <> show height
+        let imageStylePathDirectory = originalImageDirectory <> "/imageStyles/" <> size
+        let imageStylePath = imageStylePathDirectory <> "/" <> uuid
+
+        -- @todo: How to get rid of the "static/" prefix?
```
Comment (amitaibu) [inline Web/Controller/ImageStyle.hs:18]: Should we expose [storage](https://github.com/digitallyinduced/ihp/blob/248cb16b361adb4e8838b33a496f7c2be09541de/IHP/FileStorage/ControllerFunctions.hs#L363-L364) on IHP core?

File: Web/Controller/ImageStyle.hs (hunk 2 above)
Comment (mpscholten) [reply Web/Controller/ImageStyle.hs:32]: For prod usage it's btw not recommended to use the local storage. Better to use e.g. the S3 storage. The problem for production is that on NixOS the static/ directory is not writeable. So keep this in mind while desgining this controller.

File: Web/Controller/ImageStyle.hs (hunk 4 above)
Comment (mpscholten) [reply Web/Controller/ImageStyle.hs:18]: 👍

File: Web/Controller/ImageStyle.hs (hunk 2 above)
Comment (amitaibu) [reply Web/Controller/ImageStyle.hs:32]: > security reasons

Indeed, it makes the file name secure :+1: . However for my use case I need the storeFileFromPath to save the same filename, so I won't re-create the image style each time with a different name.So https://github.com/digitallyinduced/ihp/pull/1686 might not be right as is, but I think there should be a way to tell IHP the filename to save

>  Then there should be a database record that saves the storage path together with the original file name (as a string), e.g. in an attachments table.

Nice. I could add it to the docs :) . At which point do you get a hold of both filename and path?

Comment (mpscholten) [reply Web/Controller/ImageStyle.hs:32]: You could work around this by accessing the underlying `StaticDirStorage` or `S3Storage` directly, but the storage system right now is hard to move away from the UUID based approach.

> Nice. I could add it to the docs :) . At which point do you get a hold of both filename and path?

During upload, here's an example:

```haskell
module Web.Controller.UploadedFiles where

import Web.Controller.Prelude
import Web.View.UploadedFiles.Show

import qualified Network.Wai.Parse as Wai

instance Controller UploadedFilesController where
    action CreateUploadedFileAction = do
        uploadedFile <- withTransaction do 
            uploadedFile <-
                    newRecord @UploadedFile
                    |> setJust #userId currentUserId
                    |> createRecord

            let file = fileOrNothing "file" |> fromMaybe (error "no file given")

            let options :: StoreFileOptions = def
                    { directory = "uploaded_files"
                    , contentDisposition = contentDispositionAttachmentAndFileName
                    }

            storedFile <- storeFileWithOptions file options

            signedUrl <- createTemporaryDownloadUrl storedFile

            uploadedFile
                |> set #publicUrl (signedUrl |> get #url)
                |> set #publicUrlExpiredAt (signedUrl |> get #expiredAt)
                |> set #objectPath (storedFile |> get #path)
                |> set #fileName (cs $ get #fileName file)
                |> set #contentType (cs $ Wai.fileContentType file)
                |> updateRecord

        render ShowView { .. }
```

Comment (amitaibu) [reply Web/Controller/ImageStyle.hs:32]: > but the storage system right now is hard to move away from the UUID based approach.

Yeah, I'm ok with keeping the UUID. So I think we can extend [StoreFileOptions](https://github.com/digitallyinduced/ihp/blob/248cb16b361adb4e8838b33a496f7c2be09541de/IHP/FileStorage/Types.hs#L24C5-L28) to have a `fileName` property. And then I could call `storeFileFromPath` with the filename I'd like (which would be the original UUID). Sounds ok?

Comment (amitaibu) [reply Web/Controller/ImageStyle.hs:32]: I think only `storeFileWithOptions` will have to change slightly (with no BC changes). I'll create a PR to confirm

Comment (amitaibu) [reply Web/Controller/ImageStyle.hs:34]: https://github.com/digitallyinduced/ihp/pull/1688

Comment (amitaibu) [reply Web/Controller/ImageStyle.hs:34]: @mpscholten in your code example above, you are doing a few nice things that I think I could update the docs with. But I to ensure the approach:

1. If we'd have the `attachement` table it means we'd want to separate the file upload from the record build (they are [one](https://github.com/digitallyinduced/ihp/blob/f2f2af6220bdbe04df21e34b13ad561f6969d090/Guide/file-storage.markdown?plain=1#L202) now). So, like in your example we'd start with `withTransaction`, upload the file, create an `Attachement` record, and then create a `Company` record with the `logoUrl` populated.
1. We tell folks to have the `logo_url` as a [Maybe value](https://github.com/digitallyinduced/ihp/blob/f2f2af6220bdbe04df21e34b13ad561f6969d090/Guide/file-storage.markdown?plain=1#L189-L193). But what if the logoUrl should be a required field? With the above approach, we can have logoUrl be a `Text` and not `Maybe Text` -- which is better modeling the fact it's required.

Comment (amitaibu) [reply Web/Controller/ImageStyle.hs:34]: > The approach in the docs is a bit simpler.

Oh yeah, I intend to keep the simpler option. But also provide the more advanced one.

**Pull request merged**